File: 1254-number-of-closed-islands/1254-number-of-closed-islands.cpp
class Solution {
public:
    int DR[4]={0,0,1,-1};
    int DC[4]={1,-1,0,0};
    // check for a 0 , if found check if the boundary elements are not valid then dont add count
    // if the boundary elements are present then increment the count and if a 0 is found set all the nearby zeroes to 1
    void dfs(int i,int j,vector<vector<int>>& grid){
        int row=grid.size();
        int col=grid[0].size();
        if( i<0 || j<0 || i>row-1 || j>col-1 || grid[i][j]!=0) return;
        grid[i][j]=-1;
        dfs(i+1,j,grid);
        dfs(i-1,j,grid);
        dfs(i,j+1,grid);
        dfs(i,j-1,grid);
    }
    
    int closedIsland(vector<vector<int>>& grid) {
        int row = grid.size(),col = grid[0].size(),count=0; 
        for(int i =0;i<row;i++)
            for(int j=0;j<col;j++)
                if((i==0 || j==0 || i==row-1 || j==col-1) && grid[i][j]==0)
                dfs(i,j,grid);
         for(int i =0;i<row;i++)
            for(int j=0;j<col;j++)
                if(grid[i][j]==0){
                    dfs(i,j,grid);
                    count++;
                }
        return count;
    }
    /*
    void dfs(int i,int j,vector<vector<int>> &grid){
        grid[i][j]=0;
        int row = grid.size();
        int col = grid[0].size();
        for(int k=0;k<4;i++){
            int ci = DR[k]+i;
            int cj = DR[k]+j;
            if(validIndex(ci,cj,grid))  
                return 1 + dfs(ci,cj,grid);
            return 0;
        }
        
    }*/
};

File: 1254-number-of-closed-islands/1254-number-of-closed-islands.py
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        
        def dfs(x, y, m, n, visit):
            if x < 0 or x >= m or y < 0 or y >= n:
                return False
            if grid[x][y] == 1 or visit[x][y]:
                return True
            visit[x][y] = True
            isClosed = True
            for k1,k2 in [[1,0],[-1,0],[0,1],[0,-1]]:
                r = x + k1
                c = y + k2
                if not dfs(r, c, m, n, visit):
                    isClosed = False
            return isClosed

        m = len(grid)
        n = len(grid[0])
        cnt=0
        visit = [[False] * n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0 and not visit[i][j] and dfs(i,j,m,n,visit):
                    cnt+=1
        return cnt
